wpi.sh

```diff
@@ -0,0 +1,173 @@
+    exit 1
+fi
+
+# testing for JAVA11_HOME, not an unintentional reference to JAVA_HOME
+# shellcheck disable=SC2153
+if [ "x${JAVA11_HOME}" = "x" ]; then
+    echo "JAVA11_HOME must be set to a Java 11 JDK"
+    exit 1
+fi
+
+if [ ! -d "${JAVA11_HOME}" ]; then
+    echo "JAVA11_HOME is set to a non-existent directory ${JAVA11_HOME}"
+    exit 1
+fi
+
+JAVA_HOME="${JAVA11_HOME}"
+
+if [ "x${CHECKERFRAMEWORK}" = "x" ]; then
+    echo "CHECKERFRAMEWORK is not set; it must be set to a locally-built Checker Framework. Please clone and build github.com/typetools/checker-framework"
+    exit 2
+fi
+
+if [ ! -d "${CHECKERFRAMEWORK}" ]; then
+    echo "CHECKERFRAMEWORK is set to a non-existent directory ${CHECKERFRAMEWORK}"
+    exit 2
+fi
+
+if [ "x${DLJC}" = "x" ]; then
+    echo "DLJC is not set; it must be set to the dljc executable. Please checkout github.com/kelloggm/do-like-javac and point the DLJC environment variable to its dljc script"
+    exit 3
+fi
+
+if [ ! -f "${DLJC}" ]; then
+    echo "DLJC is set to a non-existent file ${DLJC}"
+    exit 2
+fi
+
+if [ ! -d "${DIR}" ]; then
+    echo "wpi.sh called on invalid directory: ${DIR}. Please supply an existing directory's absolute path."
+    exit 4
+fi
+
+function configure_and_exec_dljc {
+
+  if [ -f build.gradle ]; then
+      if [ -f gradlew ]; then
+	  chmod +x gradlew
+	  GRADLE_EXEC="./gradlew"
+      else
+	  GRADLE_EXEC="gradle"
+      fi
+      CLEAN_CMD="${GRADLE_EXEC} clean -g .gradle -Dorg.gradle.java.home=${JAVA_HOME}"
+      BUILD_CMD="${GRADLE_EXEC} clean compileJava -g .gradle -Dorg.gradle.java.home=${JAVA_HOME}"
+  elif [ -f pom.xml ]; then
+      if [ -f mvnw ]; then
+	  chmod +x mvnw
+	  MVN_EXEC="./mvnw"
+      else
+	  MVN_EXEC="mvn"
+      fi
+      # if running on java 8, need /jre at the end of this Maven command
+      if [ "${JAVA_HOME}" = "${JAVA8_HOME}" ]; then
+          CLEAN_CMD="${MVN_EXEC} clean -Djava.home=${JAVA_HOME}/jre"
+          BUILD_CMD="${MVN_EXEC} clean compile -Djava.home=${JAVA_HOME}/jre"
+      else
+          CLEAN_CMD="${MVN_EXEC} clean -Djava.home=${JAVA_HOME}"
+          BUILD_CMD="${MVN_EXEC} clean compile -Djava.home=${JAVA_HOME}"
+      fi
+  else
+      echo "no build file found for ${REPO_NAME}; not calling DLJC"
+      USABLE="no"
+      return
+  fi
+    
+  DLJC_CMD="${DLJC} -t wpi"
+
+  if [ ! "x${TIMEOUT}" = "x" ]; then
+      TMP="${DLJC_CMD}"
+      DLJC_CMD="timeout ${TIMEOUT} ${TMP}"
+  fi
+
+  TMP="${DLJC_CMD} $* -- ${BUILD_CMD}"
+  DLJC_CMD="${TMP}"
+
+  # Remove old DLJC output.
+  rm -rf dljc-out
+
+  # ensure the project is clean before invoking DLJC
+  eval "${CLEAN_CMD}" < /dev/null
+
+  echo "${DLJC_CMD}"
+
+  # This command also includes "clean"; I'm not sure why it is necessary.
+  eval "${DLJC_CMD}" < /dev/null
+
+  if [[ $? -eq 124 ]]; then
+      echo "dljc timed out for ${DIR}"
+      USABLE="no"
+      return
+  fi
+
+  if [ -f dljc-out/wpi.log ]; then
+      USABLE="yes"
```

Comment:
What does "usable" mean?  That WPI did not err?  Please clarify.
Also, this suggests that if the log file exists at all, then WPI did not err.  Is that correct?  Or does usable just mean that a log file exists?  If so, please rename the variable.

Reply:
I think `USABLE` is a fine name for this variable. If the `wpi.log` file exists, then WPI was *invoked*, so the project is usable - that is, it should be kept on disk. DLJC only actually invokes the tool (producing the associated logfile) if it is able to successfully build the project, which is really what "usable" means in this context.

Reply:
I suppose `BUILDABLE` would also be an acceptable name, but I think I still prefer `USABLE`. If you feel strongly I can change it (or, if you have another proposal, I'm happy to hear it).